File: articles/biology-scientific-understanding-living-order/python/homeostasis_model.py
"""
Homeostatic setpoint dynamics.

Run:
    python python/homeostasis_model.py
"""

from __future__ import annotations

from pathlib import Path

import numpy as np
import pandas as pd

from living_order_core import homeostatic_solution, recovery_index


ARTICLE_DIR = Path(__file__).resolve().parents[1]
HOME_PATH = ARTICLE_DIR / "data" / "homeostasis_scenarios.csv"


def main() -> None:
    """Simulate homeostatic return across scenarios."""

    scenarios = pd.read_csv(HOME_PATH)

    rows = []

    for _, scenario in scenarios.iterrows():
        time = np.arange(0, scenario["time_end"] + scenario["dt"], scenario["dt"])

        state = homeostatic_solution(
            time=time,
            initial_value=scenario["initial_value"],
            setpoint=scenario["setpoint"],
            correction_rate=scenario["correction_rate"],
        )

        rows.append(
            {
                "scenario": scenario["scenario"],
                "initial_value": scenario["initial_value"],
                "setpoint": scenario["setpoint"],
                "correction_rate": scenario["correction_rate"],
                "final_state": state[-1],
                "final_deviation": state[-1] - scenario["setpoint"],
                "half_recovery_time": (
                    np.log(2) / scenario["correction_rate"]
                    if scenario["correction_rate"] > 0
                    else np.nan
                ),
                "recovery_index": recovery_index(
                    scenario["initial_value"],
                    state[-1],
                    scenario["setpoint"],
                ),
            }
        )

    print(pd.DataFrame(rows).round(5).to_string(index=False))


if __name__ == "__main__":
    main()

File: articles/biology-scientific-understanding-living-order/r/homeostasis_model.R
# Homeostatic setpoint dynamics in R.

home_path <- file.path("data", "homeostasis_scenarios.csv")

if (!file.exists(home_path)) {
  home_path <- file.path("..", "data", "homeostasis_scenarios.csv")
}

homeostasis <- read.csv(home_path)

homeostatic_solution <- function(time, initial_value, setpoint, correction_rate) {
  setpoint + (initial_value - setpoint) * exp(-correction_rate * time)
}

recovery_index <- function(initial_value, final_value, setpoint) {
  initial_deviation <- abs(initial_value - setpoint)

  if (initial_deviation == 0) {
    return(1)
  }

  1 - abs(final_value - setpoint) / initial_deviation
}

rows <- lapply(seq_len(nrow(homeostasis)), function(i) {
  s <- homeostasis[i, ]
  time <- seq(0, s$time_end, by = s$dt)
  state <- homeostatic_solution(time, s$initial_value, s$setpoint, s$correction_rate)
  final_state <- tail(state, 1)

  data.frame(
    scenario = s$scenario,
    initial_value = s$initial_value,
    setpoint = s$setpoint,
    correction_rate = s$correction_rate,
    final_state = final_state,
    final_deviation = final_state - s$setpoint,
    half_recovery_time = ifelse(s$correction_rate > 0, log(2) / s$correction_rate, NA),
    recovery_index = recovery_index(s$initial_value, final_state, s$setpoint)
  )
})

summary_df <- do.call(rbind, rows)

print(round(summary_df, 5))
